src/tallies/tally_filter_header.F90

```diff
@@ -104,4 +129,95 @@ subroutine filter_initialize(this)
     class(TallyFilter), intent(inout) :: this
   end subroutine filter_initialize
 
+!===============================================================================
+!                               C API FUNCTIONS
+!===============================================================================
+
+  function openmc_extend_filters(n, index_start, index_end) result(err) bind(C)
+    ! Creates or extends the filters array
+    integer(C_INT32_T), value, intent(in) :: n
+    integer(C_INT32_T), optional, intent(out) :: index_start
+    integer(C_INT32_T), optional, intent(out) :: index_end
+    integer(C_INT) :: err
+
+    integer :: i ! loop counter
+    type(TallyFilterContainer), allocatable :: temp(:) ! temporary filters
+
+    if (n_filters == 0) then
+      ! Allocate filters array
+      allocate(filters(n))
+    else
+      ! Move filters to temporary array
+      allocate(temp(n_filters + n))
+      do i = 1, n_filters
+        call move_alloc(filters(i) % obj, temp(i) % obj)
+      end do
+
+      ! Move filters back from temporary array to filters array
+      call move_alloc(temp, filters)
+    end if
+
+    ! Return indices in filters array
+    if (present(index_start)) index_start = n_filters + 1
+    if (present(index_end)) index_end = n_filters + n
+    n_filters = n_filters + n
```

Comment:
Perhaps we should remove the `n_filters` variable and just use `size(filters)` to avoid inconsistent data.  Or maybe make an `n_filters()` function to handle an `if (allocated)`

Reply:
I've kind of gone back and forth on this. At one point I wanted to get rid of all the `n_` variables. However, the one nice thing about it is that you can use `n_filters` when there are zero filters, whereas `size(filters)` will cause trouble if `filters` hasn't been allocated. So, now I lean towards leaving this as is because it provides an easy way to check the size of an array from Python/ctypes.

Reply:
Good point.  What do you think of allocating arrays like this with length 0?  I think this might already be the case with tally derivatives.

Reply:
Yes, that would be another option. The only problem I see with that is that there is no way to have arrays allocated to size 0 at program startup. That could be done on the call to `openmc_init`, though. In terms of complexity, having `n_*` be functions that look at `size(...)` would probably result in more lines than the current approach. With a good API, there should only be a few spots where allocations happen (and hence where a `n_*` global variable would change).